### PR TITLE
Now we and AI can sign research agreement. AI has 50% chance to declare friendship and sign research agreement. Now i don't delete the option to trade tech.

### PR DESCRIPTION
**Pull request by lishaoxia1985**
it have two problems:
1. research agreement button can be touched when we don't have enough money to earn, in the other word, we can sign research agreement even if our gold <0.  the reason is that: in diplomacy screen it does not show the gold number of us and AI, so we don't know why we can't sign RA.
2. AI will agree to sign RA even its gold <0. 
it has two reason: the first is what the above says, the second is that i think it's a good method to consume AI's gold. what do you think?

### CODE REVIEW

**Comment by yairm210** [comment]
1. I think that RA should be disabled when we don't have enough gold, and we can communicate this clearly to the player when disabled by adding "not enough gold!" or something when disabled.
I'm confused, though, it looks like you've made the Research Agreement a trade, and not a PopupAlert?

2. I think that the AI should not be able to accept if it doesn't have enough gold, but I do agree that it's a good way to consume gold. I think what would be good is if the AI offered this themselves.

3. I'm not sure why you removed the option to trade technologies?

**Comment by lishaoxia1985** [comment]
1.    A. in different era RA costs different gold, so we don't know how much money ai has if we don't enter trade screen. so diplomat screen should be redrawed: first, split city state and civ; second, in diplomat screen,  it shows amount of money, resources in table directly and relation number,  so we can choose civ easily; third, in diplomat screen, it should only show relation number, if we want to watch serious opinion info,  we can touch a button.
        B. if we and AI are  in different era, we or Ai must pay extra gold or other things which equal amount of extra gold. using  popup alert cannot get this aim.
2.     A. the gold cost we sign with Ai  may be not equal the gold Ai signs with us if we and ai are in different era.
         B. if we have enough money but Ai does not we can give ai money to sign RA. of cource,  when we decide to sign RA we should have enough money and don't need to concern about AI's money.if both we and ai don't have enough money, sign RA will make us and AI gold＜0, science from RA will be very very little, but we and Ai have pay the money. it likes that killing 800 enemies leads to 800 people of us die. i don't know which one is right.
         C. i agree that Ai civs can sign RA with each other. but i have not find how to get this aim. your code is not relative with that they declares friendship with each other(maybe i have not found it.😀), but if they want to sign RA, they should declare friendship at first.
3.      A. if we use RA,  trade technologies should be forbidden like original game. of course, we can  use it if we make the game like civ 4.
4.     about exchange tech between AI, i know it's relative with difficulty, but we have a better choice, the below web has a table:
https://www.carlsguides.com/strategy/civilization5/difficulty-settings.php

________________________________
发件人: Yair Morgenstern <notifications@github.com>
发送时间: 2020年1月13日星期一 05:48
收件人: yairm210/Unciv
抄送: lishaoxia1985; Author
主题: Re: [yairm210/Unciv] now we can sign research agreement. (#1671)


  1.  I think that RA should be disabled when we don't have enough gold, and we can communicate this clearly to the player when disabled by adding "not enough gold!" or something when disabled.
I'm confused, though, it looks like you've made the Research Agreement a trade, and not a PopupAlert?

  2.  I think that the AI should not be able to accept if it doesn't have enough gold, but I do agree that it's a good way to consume gold. I think what would be good is if the AI offered this themselves.

  3.  I'm not sure why you removed the option to trade technologies?

**Comment by lishaoxia1985** [comment]
@yairm210 i will redraw the diplomat screen and make AI's action like original game.

**Comment by yairm210** [comment]
It's not building properly in Travis

File: core/src/com/unciv/logic/civilization/CivilizationInfo.kt [hunk 1]
```diff
@@ -495,6 +495,17 @@ class CivilizationInfo {
         updateStatsForNextTurn()
     }
 
+    fun goldCostOfSignResearchAgreement(): Int {
+        // https://forums.civfanatics.com/resources/research-agreements-bnw.25568/
+        return (when (getEra()) {
```
Comment (yairm210) [inline core/src/com/unciv/logic/civilization/CivilizationInfo.kt]: Split this into a = when{...}
return a*(...getModifier())

File: core/src/com/unciv/logic/civilization/TechManager.kt [hunk 2]
```diff
@@ -30,6 +30,7 @@ class TechManager {
     var freeTechs = 0
     /** For calculating Great Scientist yields - see https://civilization.fandom.com/wiki/Great_Scientist_(Civ5)  */
     var scienceOfLast8Turns = IntArray(8){0}
+    var getScienceFromResearchAgreement = 0
```
Comment (yairm210) [inline core/src/com/unciv/logic/civilization/TechManager.kt]: Please rename this to scienceFromResearchAgreements, 
getScience...etc is a name for a *function* that returns an amount

File: core/src/com/unciv/logic/civilization/TechManager.kt [hunk 3]
```diff
@@ -165,6 +167,16 @@ class TechManager {
...
+        if (getScienceFromResearchAgreement!=0){ // https://forums.civfanatics.com/resources/research-agreements-bnw.25568/
+            var researchAgreementModifier= 0.5f
+            if (civInfo.policies.isAdopted("Scientific Revolution"))
+                researchAgreementModifier += 0.25f
+            if (civInfo.containsBuildingUnique("Science gained from research agreements +50%"))
+                researchAgreementModifier += 0.25f
+            getScienceFromResearchAgreement = (getScienceFromResearchAgreement / 3 * researchAgreementModifier).toInt()
+            techsInProgress[currentTechnology] = techsInProgress[currentTechnology]!! + getScienceFromResearchAgreement
+            getScienceFromResearchAgreement = 0
+        }
```
Comment (yairm210) [inline core/src/com/unciv/logic/civilization/TechManager.kt]: I think this should be its own little function, getScienceFromResearchAgreements()

File: core/src/com/unciv/logic/automation/NextTurnAutomation.kt [hunk 4]
```diff
@@ -25,8 +26,9 @@ class NextTurnAutomation{
         respondToTradeRequests(civInfo)
 
         if(civInfo.isMajorCiv()) {
+            offerDeclarationOfFriendship(civInfo)
             offerPeaceTreaty(civInfo)
-            exchangeTechs(civInfo)
```
Comment (yairm210) [inline core/src/com/unciv/logic/automation/NextTurnAutomation.kt]: Why delete exchangeTechs? I think we could just move it below offerResearch

**Comment by yairm210** [comment]
Looks good in general, not sure why it's failing to build though

File: core/src/com/unciv/logic/civilization/TechManager.kt (hunk 3 above)
Comment (lishaoxia1985) [reply core/src/com/unciv/logic/civilization/TechManager.kt]: Using getScienceFromResearchAgreements() in code may lead to declaration error, so i delete the last "s". of course above this, "scienceFromResearchAgreements" does't include "s" too.

Comment (lishaoxia1985) [reply core/src/com/unciv/logic/civilization/TechManager.kt]: "getScienceFromResearchAgreements()" has the word "scienceFromResearchAgreements" in its last leads to declaration error. you can try this.

File: core/src/com/unciv/logic/civilization/CivilizationInfo.kt (hunk 1 above)
Comment (lishaoxia1985) [reply core/src/com/unciv/logic/civilization/CivilizationInfo.kt]: i will change it, but i think it has better performance.

File: core/src/com/unciv/logic/automation/NextTurnAutomation.kt (hunk 4 above)
Comment (lishaoxia1985) [reply core/src/com/unciv/logic/automation/NextTurnAutomation.kt]: exchangeTechs() has used the trade tech method, if we use exchangeTechs(), trade tech should not be removed

**Comment by lishaoxia1985** [comment]
i checkout your code not clone it,  i think it's a problem,  maybe it's about building code.
________________________________

**Comment by lishaoxia1985** [comment]
i have a suggestion:
add a setting, let player choose to trade technology or not, maybe we can make a game like civ4 to use this project.
we can make a collection game include civ4 civ5 and civ6, what do you think?
________________________________

**Comment by yairm210** [comment]
It's still failing to build...

**Comment by yairm210** [comment]
EVENTUALLY I hope that Unciv will be able to accommodate rulesets from other Civ games. But that's really, really far away...

**Comment by lyrjie** [comment]
I guess the build is failing because the added strings are missing from `English.properties`, so the translation tests fail

**Comment by lishaoxia1985** [comment]
@yairm210 @lyrjie(thank you!) I have found the reason! I will solve it. by the way, i will add a setting to make player choose trade tech.

**Comment by lishaoxia1985** [comment]
@yairm210 @lyrjie now it's ok, i will add the setting for user to choose trading tech or not.

**Comment by yairm210** [comment]
This looks ok, so I'm going to merge it now
There are things, though, that I think can be improved, so I'll comment on them =)

File: core/src/com/unciv/ui/trade/DiplomacyScreen.kt [hunk 5]
```diff
@@ -223,6 +230,38 @@ class DiplomacyScreen(val viewingCiv:CivilizationInfo):CameraStageBaseScreen() {
...
 
+            if(diplomacyManager.hasFlag(DiplomacyFlags.DeclarationOfFriendship)
+                    && !diplomacyManager.hasFlag(DiplomacyFlags.ResearchAgreement)
+                    && !otherCivDiplomacyManager.hasFlag(DiplomacyFlags.ResearchAgreement)
+                    && viewingCiv.tech.getTechUniques().contains("Enables Research agreements")
+                    && otherCiv.tech.getTechUniques().contains("Enables Research agreements")
+                    && viewingCiv.gold >= viewingCiv.goldCostOfSignResearchAgreement()
+                    && otherCiv.gold >= viewingCiv.goldCostOfSignResearchAgreement()
+                    && viewingCiv.gameInfo.ruleSet.technologies.values.any { !viewingCiv.tech.isResearched(it.name) && viewingCiv.tech.canBeResearched(it.name) }
+                    && otherCiv.gameInfo.ruleSet.technologies.values.any { !otherCiv.tech.isResearched(it.name) && otherCiv.tech.canBeResearched(it.name) }){
```
Comment (yairm210) [inline core/src/com/unciv/ui/trade/DiplomacyScreen.kt]: can be simplified to a function, canSignResearchAgreement(), and 2 calls

File: core/src/com/unciv/ui/trade/DiplomacyScreen.kt [hunk 6]
```diff
@@ -223,6 +230,38 @@ class DiplomacyScreen(val viewingCiv:CivilizationInfo):CameraStageBaseScreen() {
...
+                        GameSpeed.Epic -> 45
+                    }
+                    val researchAgreement = TradeOffer(Constants.researchAgreement, TradeType.Treaty, duration, viewingCiv.goldCostOfSignResearchAgreement())
+                    val goldCostOfSignResearchAgreement = TradeOffer("Gold".tr(), TradeType.Gold, 0, -viewingCiv.goldCostOfSignResearchAgreement())
+                    tradeTable.tradeLogic.currentTrade.theirOffers.add(researchAgreement)
+                    tradeTable.tradeLogic.ourAvailableOffers.add(researchAgreement)
+                    tradeTable.tradeLogic.ourAvailableOffers.add(goldCostOfSignResearchAgreement)
+                    tradeTable.tradeLogic.currentTrade.ourOffers.add(researchAgreement)
+                    tradeTable.tradeLogic.theirAvailableOffers.add(researchAgreement)
+                    tradeTable.tradeLogic.theirAvailableOffers.add(goldCostOfSignResearchAgreement)
```
Comment (yairm210) [inline core/src/com/unciv/ui/trade/DiplomacyScreen.kt]: more duplicates, can be simplified

File: core/src/com/unciv/logic/civilization/TechManager.kt [hunk 7]
```diff
@@ -161,11 +163,25 @@ class TechManager {
                 getRuleset().technologies[currentTechnologyName()]!!.cost))
     }
 
+    private fun scienceFromResearchAgreements(): Int {
```
Comment (yairm210) [inline core/src/com/unciv/logic/civilization/TechManager.kt]: having a parameter and a function with the exact same name can lead to confusion, please rename one of them

File: core/src/com/unciv/logic/automation/NextTurnAutomation.kt [hunk 8]
```diff
@@ -309,6 +312,62 @@ class NextTurnAutomation{
...
+        for (civ in canDeclareFriendshipCiv) {
+            // Default setting is 5, this will be changed according to different civ.
+            if ((1..10).random()<=5) civInfo.getDiplomacyManager(civ).signDeclarationOfFriendship()
+            else continue
+        }
+    }
+
+    private fun offerResearchAgreement(civInfo: CivilizationInfo) {
+        // if Civ has researched future tech, they will not want to sign RA.
+        if (civInfo.tech.getTechUniques().contains("Enables Research agreements")
```
Comment (yairm210) [inline core/src/com/unciv/logic/automation/NextTurnAutomation.kt]: It's better to reverse the 'if' to catch the negative statements and return immediately
That way we avoid indentation and stick to the leftmost 'happy path'

File: core/src/com/unciv/logic/automation/NextTurnAutomation.kt [hunk 9]
```diff
@@ -309,6 +312,62 @@ class NextTurnAutomation{
...
+                && civInfo.gold >= civInfo.goldCostOfSignResearchAgreement()
+                && civInfo.gameInfo.ruleSet.technologies.values.any { !civInfo.tech.isResearched(it.name) && civInfo.tech.canBeResearched(it.name) }){
+            val canSignResearchAgreementCiv = civInfo.getKnownCivs()
+                    .asSequence()
+                    .filter { it.isMajorCiv() }
+                    .filter { civInfo.getDiplomacyManager(it).hasFlag(DiplomacyFlags.DeclarationOfFriendship) }
```
Comment (yairm210) [inline core/src/com/unciv/logic/automation/NextTurnAutomation.kt]: if we implement the canSignResearchAgreement function, we can use it here as well and reduce code duplication

**Pull request merged**